### ISSUE TITLE
Write support

### ISSUE THREAD

**Issue by damianb** [damianb/SBAsset6#2]
Adding write support for the following libraries would be a nice to have:

* SBON
* SBAsset6

This may require some rework with ConsumableFile/ConsumableBuffer, etc.  Those libraries are designed and intended for reads *only* - not writes.

Further, with node's Buffer class being alloc-first, then write...we cannot simply stream in bytes like we would need to with SBON, which is very much a "on the fly" format.

**Comment by damianb** [comment]
Scope narrowed - this will cover just write support for SBAsset6 library.

Hinges upon the completion and availability of ExpandingBuffer and ExpandingFile.

**Comment by damianb** [comment]
Blocked by https://github.com/damianb/SBON/issues/1

**Comment by damianb** [comment]
No longer blocked, however -

Going to need to carefully consider how to do this one.  Could be a perf nightmare if incorrectly handled; editing SBAsset6 files "in place" will *not* be easy either, as we'll have to start doing some aggressive file descriptor bullshittery to try and maintain the location of the metatable.

Cases to consider:

### 1: Creating a new SBAsset6 file

Creating a new file means gathering a ton of file descriptors, reading from all of them and writing to the SBAsset6 file...then writing the metatable (this will have to come second, in order to have a file table with correct offsets), and then seeking to the beginning and updating the metatable offset.

### 2: Updating an SBAsset6 file's assets

#### (also) Removing an asset from an SBAsset6 file

We'd have to alter the specified file, track the net change in bytes, and then update the file table and all offsets (along with the specified file's length) in order to maintain accuracy of the metatable.  This could be ridiculously fragile.

It may be more effective to simply write an entirely new SBAsset6 file and just copy it over onto the old one in this case.  A lot ends up having to change or shift around...

### 3: Adding a new asset to an SBAsset6 file

We could simply dump it at the very end of the file, update the metatable as necessary, and carry on with our day (as it appears that the metatable itself is capable of living in the *middle* of the pak), but that may be deceptively simple - it appears that most Starbound-generated paks have the metatable at the *end*, which may imply that Starbound looks for it to be at the end, and the metatable offset at the beginning is simply a red herring.  If this is not a usable approach, it'll likely reuse the approach for 2.

---

The most concerning scenario, in terms of performance, is 1 - that's a lot of reading and buffering nonsense to be going through, and I bet node.js streams are going to be a mess to work with in order to get the requisite reads/writes done effectively (due to the many-read-to-one-write goal).  In terms of fragility, 2 will likely be the most dangerous.

**Comment by damianb** [comment]
Wrote StreamPipeline to handle the many-read-to-one-write situation involving streams.  Should make some things easier.
